File: src/Applications/Orbital.Web/Orbital.BlazorApp/Libraries/DynamicLinkLibraries/Motion6D/Motion6D.Interfaces/Motion6DAcceleratedFrame.cs
using System;
using System.Collections.Generic;
using System.Text;

using Vector3D;
using RealMatrixProcessor;
using Motion6D.Interfaces;

namespace Motion6D
{
    /// <summary>
    /// Accelerated reference frame
    /// </summary>
    public class Motion6DAcceleratedFrame : Motion6DFrame, IAcceleration, IAngularAcceleration
    {
        #region Fields


        double[] relativeAcceleration = new double[] { 0, 0, 0 };

        double[] acceleration = new double[] { 0, 0, 0 };

        double[] angularAcceleration = new double[] { 0, 0, 0 };

        double[] temp = new double[3];
        
        double[] tempV = new double[3];

        #endregion

        #region Overriden

        /// <summary>
        /// Sets state
        /// </summary>
        /// <param name="baseFrame">Base frame</param>
        /// <param name="relative">Relative frame</param>
        public override void Set(ReferenceFrame baseFrame, ReferenceFrame relative)
        {
            base.Set(baseFrame, relative);
            IAngularAcceleration arn = relative as IAngularAcceleration;
            IVelocity relativeVelocity = relative as IVelocity;
            IAngularVelocity baseAngulatVelocity = baseFrame as IAngularVelocity;
            IAngularVelocity relativeAngularVelocity = relative as IAngularVelocity;
            double[] rp = Position;
            double[,] m = Matrix;
            double[] relativeOmega = relativeAngularVelocity.Omega;
            double[] baseOmega = baseAngulatVelocity.Omega;
            StaticExtensionVector3D.VectorPoduct(baseOmega, relativeVelocity.Velocity, tempV);
            double om2 = StaticExtensionVector3D.Square3d(baseOmega);
            double[] eps = arn.AngularAcceleration;
            StaticExtensionVector3D.VectorPoduct(eps, rp, temp);
            for (int i = 0; i < 3; i++)
            {
                tempV[i] *= 2;
                tempV[i] += om2 * rp[i] + relativeAcceleration[i] + temp[i]; 
            }
            StaticExtensionRealMatrix.Multiply(m, tempV, acceleration);
            IOrientation relativeOrientation = relative;
            double[,] relativeMatrix = relativeOrientation.Matrix;
            StaticExtensionRealMatrix.Multiply(baseOmega, relativeMatrix, temp);
            StaticExtensionVector3D.VectorPoduct(temp, relativeOmega, tempV);
            for (int i = 0; i < 3; i++)
            {
                temp[i] = eps[i] + tempV[i];
            }
            StaticExtensionRealMatrix.Multiply(temp, m, angularAcceleration);
        }

        #endregion

        #region IAcceleration Members

        double[] IAcceleration.LinearAcceleration
        {
            get { return acceleration; }
        }

        double[] IAcceleration.RelativeAcceleration
        {
            get { return relativeAcceleration; }
        }

        #endregion

        #region IAngularAcceleration Members

        double[] IAngularAcceleration.AngularAcceleration
        {
            get { return angularAcceleration; }
        }

        #endregion
    }
}

File: src/Web/reactprojectsimplegame/src/Library/Motion6D/Motion6DAcceleratedFrame.ts
import type { IAcceleration } from "./Interfaces/IAcceleration";
import type { IAngularAcceleration } from "./Interfaces/IAngularAcceleration";
import type { IAngularVelocityMotion6D } from "./Interfaces/IAngularVelocityMotion6D";
import type { IVelocity } from "./Interfaces/IVelocity";
import { Motion6DFrame } from "./Motion6DFrame";
import { ReferenceFrame } from "./ReferenceFrame";

export class Motion6DAcceleratedFrame extends Motion6DFrame implements IAcceleration, IAngularAcceleration {
    protected relativeAcceleration: number[] = [0, 0, 0];

    protected acceleration: number[] = [0, 0, 0];

    protected angularAcceleration: number[] = [0, 0, 0];

    protected temp: number[] = [0, 0, 0];

    protected tempV: number[] = [0, 0, 0]


    constructor() {
        super();
        this.typeName = "Motion6DAcceleratedFrame";
        this.types.push("IAcceleration")
        this.types.push("IAngularAcceleration")
        this.types.push("Motion6DAcceleratedFrame")

    }


    public setReferenceFrame(baseFrame: ReferenceFrame, relative: ReferenceFrame): void {
        super.setReferenceFrame(baseFrame, relative)
        let arn = this.performer.convertObject<IAngularAcceleration, ReferenceFrame>(relative, " IAngularAcceleration")
        let relativeVelocity = this.performer.convertObject<IVelocity, ReferenceFrame>(relative, "IVelocity")
        let baseAngulatVelocity = this.performer.convertObject<IAngularVelocityMotion6D, ReferenceFrame>(baseFrame, "IAngularVelocityMotion6D")
        let relativeAngularVelocity = this.performer.convertObject<IAngularVelocityMotion6D, ReferenceFrame>(relative, "IAngularVelocityMotion6D")
        var rp = this.getPosition();
        let m = this.getMatrix();
        let relativeOmega = relativeAngularVelocity[0].getOmega();
        let baseOmega = baseAngulatVelocity[0].getOmega();
        this.vp.vectorProduct(baseOmega, relativeVelocity[0].getVelocity(), this.tempV);
        let om2 = this.vp.square3d(baseOmega);
        let eps = arn[0].getAngularAcceleration();
        this.vp.vectorProduct(eps, rp, this.temp);
        for (let i = 0; i < 3; i++) {
            this.tempV[i] *= 2;
            this.tempV[i] += om2 * rp[i] + this.relativeAcceleration[i] + this.temp[i];
        }
        this.realMatrix.multiplyRight(m, this.tempV, this.acceleration);
        var relativeOrientation = relative;
        var relativeMatrix = relativeOrientation.getMatrix();
        this.realMatrix.multiplyLeft(baseOmega, relativeMatrix, this.temp);
        this.vp.vectorProduct(this.temp, relativeOmega, this.tempV);
        for (let i = 0; i < 3; i++) {
            this.temp[i] = eps[i] + this.tempV[i];
        }
        this.realMatrix.multiplyLeft(this.temp, m, this.angularAcceleration);

    }


    getAngularAcceleration(): number[] {
        return this.angularAcceleration;
    }
    getLineraAcceleration(): number[] {
        return this.acceleration; ;
    }
    getRelativeAcceleration(): number[] {
        return this.relativeAcceleration;
    }


}